File: packages/gheim-js/src/detectors/remote.ts
/** Remote detector — calls a gheim-server (or compatible) over HTTP. */
import { type Detector, type Span, trimSpan } from "./base.ts";

export const DEFAULT_BASE_URL = "https://api.gheim.ch";
export const DEFAULT_MODEL = "openai/privacy-filter";

export interface RemoteDetectorOptions {
  baseUrl?: string;
  apiKey?: string;
  model?: string;
  /** Strip leading/trailing whitespace from each span (default: true). */
  trimWhitespace?: boolean;
  fetch?: typeof fetch;
}

export class RemoteDetector implements Detector {
  private readonly baseUrl: string;
  private readonly apiKey?: string;
  private readonly model: string;
  private readonly trimWhitespace: boolean;
  private readonly fetchImpl: typeof fetch;

  constructor(opts: RemoteDetectorOptions = {}) {
    const envBase = typeof process !== "undefined" ? process.env?.GHEIM_BASE_URL : undefined;
    const envKey = typeof process !== "undefined" ? process.env?.GHEIM_API_KEY : undefined;
    this.baseUrl = (opts.baseUrl ?? envBase ?? DEFAULT_BASE_URL).replace(/\/+$/, "");
    this.apiKey = opts.apiKey ?? envKey;
    this.model = opts.model ?? DEFAULT_MODEL;
    this.trimWhitespace = opts.trimWhitespace ?? true;
    this.fetchImpl = opts.fetch ?? fetch;
  }

  async detect(text: string, opts?: { model?: string }): Promise<Span[]> {
    if (!text) return [];
    const headers: Record<string, string> = { "Content-Type": "application/json" };
    if (this.apiKey) headers.Authorization = `Bearer ${this.apiKey}`;
    const resp = await this.fetchImpl(`${this.baseUrl}/v1/detect`, {
      method: "POST",
      headers,
      body: JSON.stringify({ model: opts?.model ?? this.model, text }),
    });
    if (!resp.ok) {
      throw new Error(`gheim-server ${resp.status}: ${await resp.text()}`);
    }
    const data = (await resp.json()) as { spans?: Array<Omit<Span, "text">> };
    const out: Span[] = [];
    for (const item of data.spans ?? []) {
      let { start, end } = item;
      if (this.trimWhitespace) {
        const trimmed = trimSpan(text, start, end);
        if (!trimmed) continue;
        ({ start, end } = trimmed);
      }
      out.push({
        label: item.label,
        start,
        end,
        text: text.slice(start, end),
        score: item.score,
      });
    }
    return out;
  }
}

File: packages/gheim-py/src/gheim/detectors/remote.py
"""Remote detector backend — calls a gheim-server (or compatible) over HTTP."""
from __future__ import annotations

import os
from dataclasses import dataclass, field
from typing import Annotated, Any

import httpx

from .base import Span, trim_span

DEFAULT_BASE_URL = "https://api.gheim.ch"
DEFAULT_MODEL = "openai/privacy-filter"


@dataclass
class RemoteDetector:
    """POSTs ``text`` to ``{base_url}/v1/detect`` and returns spans.

    Wire protocol:
        request:  { "model": "...", "text": "..." }
        response: { "spans": [ { "label", "start", "end", "score" }, ... ] }

    The server only ever sees the input text and returns spans — never the
    LLM-bound traffic, never the eventual response.
    """

    base_url: Annotated[str | None, "Base URL of the detection server. Falls back to GHEIM_BASE_URL."] = None
    api_key: Annotated[str | None, "Bearer token. Falls back to GHEIM_API_KEY."] = None
    model: Annotated[str, "Default model id passed in the request body."] = DEFAULT_MODEL
    timeout: Annotated[float, "Per-request timeout in seconds."] = 30.0
    trim_whitespace: Annotated[
        bool, "Strip leading/trailing whitespace from each span (BPE artifact fix)."
    ] = True
    client: Annotated[httpx.Client | None, "Pre-built httpx client; created on demand if None."] = None
    _owns_client: bool = field(default=False, init=False, repr=False)

    def __post_init__(self) -> None:
        self.base_url = (
            self.base_url or os.getenv("GHEIM_BASE_URL") or DEFAULT_BASE_URL
        ).rstrip("/")
        if self.api_key is None:
            self.api_key = os.getenv("GHEIM_API_KEY")
        self._owns_client = self.client is None

    def _http(self) -> httpx.Client:
        if self.client is None:
            self.client = httpx.Client(
                base_url=self.base_url or DEFAULT_BASE_URL,
                timeout=self.timeout,
            )
        return self.client

    def _headers(self) -> dict[str, str]:
        return {"Authorization": f"Bearer {self.api_key}"} if self.api_key else {}

    def detect(
        self,
        text: Annotated[str, "Source text to scan for PII."],
        *,
        model: Annotated[str | None, "Per-call model id override."] = None,
    ) -> list[Span]:
        if not text:
            return []
        payload = {"model": model or self.model, "text": text}
        resp = self._http().post("/v1/detect", json=payload, headers=self._headers())
        resp.raise_for_status()
        data: dict[str, Any] = resp.json()
        spans: list[Span] = []
        for item in data.get("spans", []):
            start = int(item["start"])
            end = int(item["end"])
            if self.trim_whitespace:
                trimmed = trim_span(text, start, end)
                if trimmed is None:
                    continue
                start, end = trimmed
            spans.append(
                Span(
                    label=str(item["label"]),
                    start=start,
                    end=end,
                    text=text[start:end],
                    score=float(item.get("score", 1.0)),
                )
            )
        return spans

    def close(self) -> None:
        if self.client is not None and self._owns_client:
            self.client.close()
            self.client = None

    def __enter__(self) -> RemoteDetector:
        return self

    def __exit__(self, *exc: Any) -> None:
        self.close()
